x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/ChunkedTrainedModelRestorer.java

```diff
@@ -201,6 +214,31 @@ private void doSearch(
         }
     }
 
+    static SearchResponse retryingSearch(Client client, String modelId, SearchRequest searchRequest, int retries, TimeValue sleep)
+        throws InterruptedException {
+        int failureCount = 0;
+
+        while (true) {
+            try {
+                return client.search(searchRequest).actionGet();
+            } catch (Exception e) {
+                if (ExceptionsHelper.unwrapCause(e) instanceof SearchPhaseExecutionException == false
+                    && ExceptionsHelper.unwrapCause(e) instanceof CircuitBreakingException == false) {
+                    throw e;
+                }
+
+                if (failureCount >= retries) {
+                    logger.warn(format("[%s] searching for model part failed %s times, returning failure", modelId, retries));
+                    throw e;
```

Comment:
Should the retry logic in `TrainedModelAssignmentNodeService` now be removed? If the search has failed after N tries it trying to reload the entire model puts the cluster under more strain.

https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentNodeService.java#L219

Instead of throwing `e` please wrap it in another exception with a sensible error message saying something along the lines of `loading model [model_id] failed after [SEARCH_RETRY_LIMIT] retries. The deployment is now in a failed state, the error may be transient please stop the deployment and restart`.

The error message ends up in the failed routing state here and will eventually make it's way back to the caller assuming the request did not time out.

https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentNodeService.java#L718

Reply:
Ah good point 👍

Reply:
Dave and I talked about this offline and we don't need to adjust the retry logic in `TrainedModelAssignmentNodeService` because using `ElasticsearchException` cannot be unwrapped and will not trigger that retry logic.